mopidy_spotify/utils.py

```diff
@@ -33,3 +35,71 @@ def time_logger(name, level=TRACE):
 
 def flatten(list_of_lists):
     return [item for sublist in list_of_lists for item in sublist]
+
+
+class op:
+    def __init__(self, op, tracks, frm, to=0):
+        self.op = op
+        self.tracks = tracks
+        self.frm = frm
+        self.to = to
+
+    def __repr__(self):
+        length = len(self.tracks)
+        if length < 1:
+            return "<invalid op()>"
+        first = self.tracks[0].split(":")[-1]
+        last = self.tracks[-1].split(":")[-1]
+        tracks = f"{first}...{last}" if length > 1 else first
+        action = {"-": "delete", "+": "insert", "m": "move"}
+        pos = f"{self.frm} to {self.to}" if self.op == "m" else self.frm
+        return f"<{action.get(self.op)} {length} tracks [{tracks}] at {pos}>"
+
+
+def _is_move(op1, op2):
+    return op1.op == "-" and op2.op == "+" and op1.tracks == op2.tracks
+
+
+def _op_split(o, chunksize):
+    def partition(o, n):
+        def inc(m, i):
+            return m + i if o.op == "-" or o.op == "m" else m
+
+        for i in range(0, len(o.tracks), n):
+            yield op(o.op, o.tracks[i : i + n], inc(o.frm, i), inc(o.to, i))
+
+    return list(partition(o, chunksize)) if o.op != "m" else [o]
+
+
+def diff(old, new, chunksize=100):
+    # first, apply python's built-in diff algorithm, remove unmodified ranges,
+    # split replacements into seperate insertions and deletions and transform
+    # the data structure into op-objects:
+    ops = []
+    for tag, i1, i2, j1, j2 in difflib.SequenceMatcher(
+        a=old, b=new, autojunk=False
+    ).get_opcodes():
+        if tag in ("insert", "replace"):
```

Comment:
Do I understand it correctly that for a real replace operation (i.e. NOT a move) that would cost one API request, it's still converted into an add/delete pair that costs two API requests. Is there some situation where actually they can be merged into a move, or is this just a simplification? The latter is fine for now, just trying to understand. For example:
```
old = ['0', '1', '2', '3']
new = ['0', '1', 'A', 'B']
difflib.SequenceMatcher(a=old, b=new, autojunk=False).get_opcodes()
# [('equal', 0, 2, 0, 2), ('replace', 2, 4, 2, 4)]
# Becomes:
# [<insert 2 tracks [A...B] at 2>, <delete 2 tracks [2...3] at 2>]
 

```

Reply:
not exactly sure what you're getting at. a "replace" operation always removes and adds some tracks, so it'll always result in two api calls. merging into mov operations happens afterwards, in the code block below.

Reply:
splitting "replace"s into ins/del pairs is necessary to catch the most number of movs. for example, `difflib.SequenceMatcher(a="12xy34ab", b="12ab34", autojunk=False).get_opcodes()` generates a replace and a delete; this will turn into an insert, delete, delete, and then below in a mov, delete.

Reply:
Apologies, I invented a new Web API request method that did replacing! It's been one of those days....

Reply:
no worries, happens to the best of us ;)